lib/gemini.c

```diff
@@ -0,0 +1,267 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 1998 - 2020, Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#if !defined CURL_DISABLE_GEMINI && defined USE_SSL
+
+#include <ctype.h>
+#include <string.h>
+#include "gemini.h"
+#include "urldata.h"
+#include "vtls/vtls.h"
+#include <curl/curl.h>
+#include "transfer.h"
+#include "sendf.h"
+#include "connect.h"
+#include "multiif.h"
+#include "progress.h"
+#include "gopher.h"
+#include "select.h"
+#include "strdup.h"
+#include "url.h"
+#include "escape.h"
+#include "warnless.h"
+#include "curl_printf.h"
+#include "curl_memory.h"
+/* The last #include file should be: */
+#include "memdebug.h"
+
+static CURLcode gemini_setup_connection(struct connectdata *conn)
+{
+  struct GEMINI *gemini;
+  struct Curl_easy *data = conn->data;
+  DEBUGASSERT(data->req.p.gemini == NULL);
+
+  gemini = calloc(1, sizeof(struct GEMINI));
+  if(!gemini)
+    return CURLE_OUT_OF_MEMORY;
+  data->req.p.gemini = gemini;
+  return CURLE_OK;
+}
+
+static CURLcode gemini_connecting(struct connectdata *conn, bool *done)
+{
+  return Curl_ssl_connect_nonblocking(conn, FIRSTSOCKET, done);
+}
+
+static CURLcode gemini_do_it(struct connectdata *conn, bool *done)
+{
+  struct GEMINI *gemini;
+  struct Curl_easy *data;
+  char *request;
+
+  data = conn->data;
+  request = aprintf("%s\r\n", data->change.url);
+
+  if(!request)
+    return CURLE_OUT_OF_MEMORY;
+
+  gemini = data->req.p.gemini;
+  gemini->request.memory = request;
+  gemini->request.amount_total = strlen(request);
+  gemini->request.amount_sent = 0;
+
+  /* Real work happens in gemini_doing, so we can use non-blocking
+   * functions and avoid busy loops.
+   */
+
+  return CURLE_OK;
+}
+
+static CURLcode gemini_doing_finish(struct connectdata *, bool *);
+static CURLcode gemini_doing(struct connectdata *conn, bool *done)
+{
+  CURLcode err;
```

Comment:
If you'd make the `CURLcode' variables be named 'result' you would follow standard curl code patterns closer. Not a requirement, just an idea.

Reply:
Renamed.